core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -143,49 +148,108 @@ static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
 	io_write64(qm->io_base + QM_DOORBELL_SQ_CQ_BASE, doorbell);
 }
 
-static enum hisi_drv_status qm_wait_mb_ready(struct hisi_qm *qm)
+static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
 {
-	uint32_t val = 0;
-
-	/* Return 0 mailbox ready, HISI_QM_DRVCRYPT_ETMOUT hardware timeout */
-	if (IO_READ32_POLL_TIMEOUT(qm->io_base + QM_MAILBOX_BASE, val,
-				   !(val & QM_MB_BUSY_BIT), POLL_PERIOD,
-				   POLL_TIMEOUT)) {
-		return HISI_QM_DRVCRYPT_ETMOUT;
-	}
+	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
 
-	return HISI_QM_DRVCRYPT_NO_ERR;
+	write_64bit_pair(dst, mb->x[0], mb->x[1]);
 }
 
-static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
+static void qm_mb_read(struct hisi_qm *qm, struct qm_mailbox *mb)
 {
-	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
+	vaddr_t mb_base = qm->io_base + QM_MAILBOX_BASE;
 
-	write_64bit_pair(dst, mb->x[0], mb->x[1]);
+	write_64bit_pair((vaddr_t)mb, mb_base, mb_base + U64_OFFSET);
```

Comment:
Aren't you supposed to read 16 bytes from `qm->io_base + QM_MAILBOX_BASE`?

Is this tested?

Reply:
![image](https://github.com/OP-TEE/optee_os/assets/118411745/14d27eb2-6d57-4152-a599-bdf5f4896e3f)
The original code was implemented this way. Referring to qm_mb_read, I have re implemented qm_mb_write. Due to significant differences between code on our library and code in community, it is difficulet to test the modified code, but the original code has been tested.I am sorry about that.I will open up the remaining patches as soon as possible and then turn to the community's code.

Reply:
This is confusing. Why are you using `write_64bit_pair()` to do an atomic 128-bit write into the mailbox which isn't shared with the device? It seems a plain `memcpy()` or such would do the job just as well.

Reply:
The last two arguments to `write_64bit_pair()` are addresses, while the actual data is expected. I don't think this function works as intended.

Reply:
The hardware requires that 128bits should be read from hardware at one time.It is implemented in kernel in that way using stp as picture above.My modification was wrong.Solved now.